src/photos/ducks/timeline-clusters/index.js

```diff
@@ -36,39 +38,122 @@ const TIMELINE_MUTATIONS = query => ({
+    } else {
+      // If the section has no period, it is not a cluster but a daily section
+      return differenceInCalendarDays(datetime, new Date(date)) === 0
+    }
+  })
+}
+
+/**
+ * Save the photos not clustered in the sections.
+ * It simulates a clustering, waiting for the actual one to be processed.
+ */
+const handlePhotosNotClustered = (sections, photosNotClustered) => {
+  photosNotClustered.forEach(photo => {
     const datetime =
-      p.metadata && p.metadata.datetime ? p.metadata.datetime : Date.now()
-    // here we want to get an object whose keys are months in a l10able format
-    // so we only keep the year and month part of the date
-    const month = datetime.slice(0, 7) + '-01T00:00'
-    /* istanbul ignore else */
-    if (!sections.hasOwnProperty(month)) {
-      sections[month] = []
+      photo.metadata && photo.metadata.datetime
+        ? photo.metadata.datetime
+        : photo.created_at
+
+    const sectionDate = matchingSection(sections, new Date(datetime))
+    if (sectionDate) {
+      sections[sectionDate].photos.push(photo)
+    } else {
+      // Create a new section for this day, without a period, to differentiate from clusters' sections
+      const day = datetime.slice(0, 10)
+      sections[day] = {
+        date: day,
+        title: formatDate(new Date(datetime)),
+        photos: [photo]
+      }
+    }
+  })
+}
+
+/**
+ * Retrieve photos by clusters. Each clusterized photo is referenced by an album
+ * having an 'auto' field to true.
+ * If a photo is not clusterized yet (newly updated ones), we try to insert them
+ * inside existing clusters, or group them per day otherwise.
+ */
+const getPhotosByClusters = photos => {
+  const sections = {}
+  const photosNotClustered = []
+
+  photos.forEach(p => {
+    const refAlbums = p.albums ? p.albums.data : null
+    // TODO Ensure unicity on the service side
+    const album = refAlbums ? refAlbums.find(ref => ref.auto) : null
+
+    // The photo is not referenced by an auto album yet
+    if (!album) {
+      photosNotClustered.push(p)
+    } else {
+      const date = album.name
+
+      if (!sections.hasOwnProperty(date)) {
+        const title = sectionTitle(album)
+        sections[date] = {
+          photos: [],
+          title: title,
+          period: album.period
+        }
+      }
+      sections[date].photos.push(p)
     }
-    sections[month].push(p)
   })
-  // we need to sort the months here because when new photos are uploaded, they
-  // are inserted on top of the list, and months can become unordered
-  const sortedMonths = Object.keys(sections)
-    .sort((a, b) => new Date(a).getTime() - new Date(b).getTime())
-    .reverse()
 
-  return sortedMonths.map(month => {
+  // We deal with the not-clustered photos after the loop to make sure all
+  // the sections have been processed.
+  if (photosNotClustered.length > 0) {
+    handlePhotosNotClustered(sections, photosNotClustered)
```

Comment:
I think it's a bit dangerous to pass around the `sections` object, as it's a critical variable. 
For example in `handlePhotosNotClustered `, we add new sections, but those section have a `date` property that other sections don't have and they don't have a period. Having these slight differences scattered across multiple functions makes them hard to find, and changing the structure tricky. There's also the fact that `period` seems to be only used for the non-custered part of the code whereas `title` and `photos` are important afterwards.

 I'm not quite sure what the best approach is here — inlining `handlePhotosNotClustered` might be a good start though.